school/module.py

```diff
@@ -0,0 +1,1499 @@
+
+        return teachers
+
+    async def _import_programs(self, ctx, json_data, degree):
+        """Import programs from JSON data"""
+        programs = []
+        for program_data in json_data:
+            spec_degree = re.search(r"^[A-z]-", program_data)
+            if spec_degree:
+                spec_degree = spec_degree.group()
+                if spec_degree[0] in DEGREE_MAPPING:
+                    spec_degree = DEGREE_MAPPING[spec_degree[0]]
+                else:
+                    await guild_log.warning(
+                        ctx.author,
+                        ctx.channel,
+                        f'Could not find mapping for spec_degree "{spec_degree}".',
+                    )
+                    spec_degree += "UNKNOWN"
+                program_data = re.sub(r"^[A-z]-", "", program_data)
+            program_info = program_data.split("-")
+            programs.append(
+                {
+                    "abbreviation": program_info[0]
+                    if program_info[1] != "A"
+                    else program_info[0] + "-" + program_info[1],
+                    "year": program_info[-2],
+                    "obligation": program_info[-1],
+                    "degree": spec_degree if spec_degree else degree,
+                }
+            )
+
+        return programs
+
+    async def _import_school_data(self, ctx, json_data) -> int:
+        """Main logic for importing school (subject) data from JSON"""
+        i = 0
+        for subject_data in json_data:
+            guarantor_data = subject_data.get("guarantor", {})
+            teachers_data = subject_data.get("teachers", {})
+            programs_data = subject_data.get("programmes", [])
+            bachelors_degree = subject_data.get("bachelors_degree", False)
+            masters_degree = subject_data.get("masters_degree", False)
+            doctoral_degree = subject_data.get("doctoral_degree", False)
+
+            if bachelors_degree:
+                degree = "Bachelor"
+            elif masters_degree:
+                degree = "Masters"
+            elif doctoral_degree:
+                degree = "Doctoral"
+            else:
+                degree = None
+
+            guarantor = await self._import_teachers(ctx, guarantor_data)
+            teachers = await self._import_teachers(ctx, teachers_data)
+
+            subject = Subject.from_json(ctx, subject_data)
+
+            if not subject:
+                await guild_log.warning(
+                    ctx.author,
+                    ctx.channel,
+                    f'Subject could not be created. Data: "{subject_data}"',
+                )
+                continue
+
+            subject.set_guarantor(guarantor)
+            subject.set_teachers(teachers)
+
+            programs = await self._import_programs(ctx, programs_data, degree)
+
+            subject.import_programs(ctx, programs)
+            i += 1
+
+        return i
```

Comment:
OK, let's see how I'd handle this:

1. You delete all Subjects that don't have a review. Same with Teachers
2. You delete all Programs, SubjectPrograms and SubjectUrls.
3. You loop through the new program data and add each to the DB.
4. You loop through the new subject data and for each:
    1. Add the subject to the DB
    2. You create the SubjectPrograms relationships
    3. You add Teachers to the DB from both the guarantor and teacher fields
    4. You create the TeacherSubject relationships
    5. You create the SubjectUrls and add the relationships

There, now all the data is up to date and good to use. No need for admin commands, no need for added complexity inside the database file.

Reply:
I'd like not to delete anything during import. That should be handled by purge command that purges unused data (not reviewed or not updated since some date)

Reply:
But how would you handle conflicts during import? If you don't enforce the deletion of unused data, you are risking duplicate entries (subjects abbreviation changes because there's a new acreditation for example) or otherwise weird stuff happenning